File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/move/filter/BoardCycleFilter.java
package org.silnith.game.solitaire.move.filter;

import java.util.Iterator;
import java.util.List;

import org.silnith.game.GameState;
import org.silnith.game.solitaire.Board;
import org.silnith.game.solitaire.move.SolitaireMove;

/**
 * Filters moves that introduce a cycle into the board history.
 */
public class BoardCycleFilter implements SolitaireMoveFilter {

    @Override
    public Object getStatisticsKey() {
        return "Board Cycle";
    }

    @Override
    public boolean shouldFilter(final List<GameState<SolitaireMove, Board>> gameStateHistory) {
        assert gameStateHistory != null;

        final Iterator<GameState<SolitaireMove, Board>> iterator = gameStateHistory.iterator();
        assert iterator.hasNext();
        final GameState<SolitaireMove, Board> currentGameState = iterator.next();
        final Board currentBoard = currentGameState.getBoard();

        while (iterator.hasNext()) {
            final GameState<SolitaireMove, Board> gameState = iterator.next();
            if (currentBoard.equals(gameState.getBoard())) {
                return true;
            }
        }
        return false;
    }

}

File: cpp/Solitaire/Klondike/silnith/game/solitaire/move/filter/BoardCycleFilter.cpp
#include <silnith/game/solitaire/move/filter/BoardCycleFilter.h>

using namespace std;
using namespace std::literals::string_literals;

namespace silnith::game::solitaire::move::filter
{
    string BoardCycleFilter::get_statistics_key() const
    {
        return "Board Cycle"s;
    }

    bool BoardCycleFilter::should_filter(shared_ptr<linked_node<game_state<solitaire_move, board>> const> const& game_state_history) const
    {
        linked_node<game_state<solitaire_move, board>>::const_iterator iterator{ game_state_history->cbegin() };
        linked_node<game_state<solitaire_move, board>>::const_iterator end{ game_state_history->cend() };
        game_state<solitaire_move, board> const& current_game_state{ *iterator };
        iterator++;
        shared_ptr<board const> const& current_board{ current_game_state.get_board() };

        while (iterator != end)
        {
            game_state<solitaire_move, board> const& game_state{ *iterator };
            iterator++;
            shared_ptr<board const> const& past_board{ game_state.get_board() };
            if (*current_board == *past_board)
            {
                return true;
            }
        }
        return false;
    }
}
